File: LC/Python/Medium/11.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        l, r = 0, len(height) - 1 
        ans = 0
        min_h = 0
        
        while l < r:
            min_h = min(height[l], height[r])
            area = min_h * (r - l)
            ans = max(ans, area)
            
            while l < r and height[l] <= min_h:
                l += 1
            while l < r and height[r] <= min_h: 
                r -= 1
            
        return ans

File: LC/C++/Medium/11.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        int n = height.size();
        int l = 0;
        int r = n - 1;
        int ans = 0;

        while (l < r) {
            int curr_h = min(height[l], height[r]);
            int curr_w = r - l;
            ans = max(ans, curr_h * curr_w);

            if (height[l] < height[r]) {
                l += 1;
            } else {
                r -= 1;
            }

        }

        return ans;
    }
};